engine/src/main/java/org/apache/apex/engine/plugin/DefaultApexPluginDispatcher.java

```diff
@@ -0,0 +1,123 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.apex.engine.plugin;
+
+import java.util.NoSuchElementException;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.RejectedExecutionHandler;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.apex.engine.api.DAGExecutionPluginContext.Handler;
+import org.apache.apex.engine.api.DAGExecutionPluginContext.RegistrationType;
+import org.apache.apex.engine.api.PluginLocator;
+import org.apache.hadoop.conf.Configuration;
+
+import com.datatorrent.common.util.NameableThreadFactory;
+import com.datatorrent.stram.StramAppContext;
+import com.datatorrent.stram.StreamingContainerManager;
+import com.datatorrent.stram.webapp.AppInfo;
+
+/**
+ * Handle dispatching of events from Stram to Plugins. This implementation creates an executor
+ * service to process the event asynchronously. A separate task {@link DefaultApexPluginDispatcher.ProcessEventTask}
```

Comment:
@tweise I also incorrectly assumed that your comment was about dispatching of the events. The access to data provided through context is thread safe except the DAG. The other methods are already used from different thread such as webservice and appdata push agent. For DAG we could

- Remove DAG access and provide explicit DAG change event
- Wrap DAG access through Read-only dag object which prevent DAG modifications but allow access to DAG elements. (https://github.com/tushargosavi/apex-core/blob/dag_schedule.redesign.1/engine/src/main/java/com/datatorrent/stram/plan/logical/ReadOnlyLogicalPlan.java)
- Use  snapshot of DAG.